File: Engine/Source/Runtime/Core/Private/Logging/Log.cpp
// Copyright 2020-2022 Aumoa.lib. All right reserved.

#include "Logging/Log.h"
#include "Logging/LogCategory.h"
#include "Logging/LogEntry.h"
#include "Logging/ILoggingExtension.h"
#include "Threading/Spinlock.h"
#include "Threading/SpinlockConditionVariable.h"
#include "Misc/PlatformMisc.h"
#include "Console.h"
#include <future>

namespace
{
	static std::jthread sWorker;
	static bool sRunning;
	static Spinlock sLock;
	static SpinlockConditionVariable sCondVar;
	static std::vector<LogEntry> sEntries;
	static std::vector<std::function<void()>> sActions;
	static std::vector<std::unique_ptr<ILoggingExtension>> sExtensions;
}

void Log::Initialize()
{
	sRunning = true;
	sWorker = std::jthread(_worker);
}

void Log::Print(const LogCategory& logCategory, ELogLevel logLevel, const String& message)
{
	static int trap = (_trap_init(), 0);

	std::unique_lock lock(sLock);
	sEntries.emplace_back() =
	{
		.LogDate = DateTime::Now(),
		.Category = &logCategory,
		.LogThread = _get_threadid_or_name(),
		.Level = logLevel,
		.Message = message
	};
	sCondVar.NotifyOne();

	if (logLevel == ELogLevel::Fatal)
	{
		lock.unlock();
		FlushAll();
	}
}

void Log::Post(std::function<void()> work)
{
	static int trap = (_trap_init(), 0);

	std::unique_lock lock(sLock);
	sActions.emplace_back(work);
	sCondVar.NotifyOne();
}

void Log::FlushAll()
{
	static int trap = (_trap_init(), 0);

	std::promise<void> promise;
	std::future<void> future = promise.get_future();

	std::unique_lock lock(sLock);
	sActions.emplace_back([&]()
	{
		promise.set_value();
	});
	sCondVar.NotifyOne();

	lock.unlock();
	future.wait();
}

void Log::InstallExtension(std::unique_ptr<ILoggingExtension> extension)
{
	extension->Installed();

	auto lock = std::unique_lock(sLock);
	sExtensions.emplace_back(std::move(extension));
}

void Log::Cleanup()
{
	static int trap = (_trap_init(), 0);

	auto lock = std::unique_lock(sLock);
	sRunning = false;
	sCondVar.NotifyOne();
	lock.unlock();

	sWorker.join();

	for (auto& extension : sExtensions)
	{
		extension->Uninstalled();
	}

	sExtensions.clear();
}

String Log::_get_threadid_or_name() noexcept
{
	static thread_local String name = []()
	{
		String name = PlatformMisc::GetThreadName();
		if (name.length() == 0)
		{
			name = String::Format(TEXT("Thread {}"), PlatformMisc::GetCurrentThreadId());
		}
		return name;
	}();
	return name;
}

void Log::_trap_init()
{
	struct _cleanup
	{
		~_cleanup()
		{
			Log::Cleanup();
		}
	};

	static auto _s = (Initialize(), _cleanup());
}

void Log::_worker()
{
	std::vector<LogEntry> entries;
	std::vector<std::function<void()>> actions;

	static auto levelToString = [](ELogLevel level) -> String
	{
		switch (level)
		{
		case ELogLevel::Debug:
			return TEXT("dbug");
		case ELogLevel::Info:
			return TEXT("info");
		case ELogLevel::Warning:
			return TEXT("warn");
		case ELogLevel::Error:
			return TEXT("erro");
		default:
			return TEXT("ftal");
		}
	};

	static auto levelToColor = [](ELogLevel level) -> EConsoleColor
	{
		switch (level)
		{
		case ELogLevel::Debug:
			return EConsoleColor::White;
		case ELogLevel::Info:
			return EConsoleColor::DarkGreen;
		case ELogLevel::Warning:
			return EConsoleColor::DarkYellow;
		case ELogLevel::Error:
			return EConsoleColor::DarkRed;
		default:
			return EConsoleColor::Red;
		}
	};

	static auto previousColor = Console::GetForegroundColor();

	while (sRunning || !sEntries.empty() || !sActions.empty())
	{
		std::unique_lock lock(sLock);
		std::swap(entries, sEntries);
		std::swap(actions, sActions);
		lock.unlock();

		for (auto& entry : entries)
		{
			Console::SetForegroundColor(levelToColor(entry.Level));
			Console::Write(levelToString(entry.Level));
#if !SHIPPING
			PlatformMisc::OutputDebugString(levelToString(entry.Level));
#endif
			Console::SetForegroundColor(previousColor);

			String formatted;
			formatted += String::Format(TEXT(": {}[{}]\n"), entry.Category->GetCategoryName(), entry.LogThread);
			for (auto& line : entry.Message.Split('\n'))
			{
				formatted += String::Format(TEXT("      {}\n"), line.TrimStart());
			}
			for (auto& extension : sExtensions)
			{
				extension->TraceOne(entry);
			}

			Console::Write(formatted);
#if !SHIPPING
			PlatformMisc::OutputDebugString(formatted);
#endif
		}
		for (auto& extension : sExtensions)
		{
			extension->Flush();
		}

		for (auto& action : actions)
		{
			action();
		}

		entries.clear();
		actions.clear();
		lock.lock();
		sCondVar.Wait(lock, []()
		{
			return !sEntries.empty() || !sRunning || !sActions.empty();
		});
	}
}

File: Engine/Source/Programs/DotNETUtilities/Diagnostics/Log.cs
// Copyright 2020-2022 Aumoa.lib. All right reserved.

namespace AE.Diagnostics;

public static class Log
{
    private static readonly Thread LogThread;
    private static readonly CancellationTokenSource StopToken = new();
    private static List<LogEntry> LogEntries = new();

    static Log()
    {
        LogThread = new Thread(Run);
        LogThread.Start();
    }

    public static void Verbose(LogVerbosity Verboisty, string Format, params object?[] Args)
    {
        lock (LogThread)
        {
            LogEntries.Add(new LogEntry
            {
                Verbosity = Verboisty,
                Message = string.Format(Format, Args)
            });

            Monitor.Pulse(LogThread);
        }
    }

    public static void Action(Action LogAction)
    {
        lock (LogThread)
        {
            LogEntries.Add(new LogEntry
            {
                LogAction = LogAction
            });

            Monitor.Pulse(LogThread);
        }
    }

    public static void Trace(string Format, params object?[] Args) => Verbose(LogVerbosity.Trace, Format, Args);

    public static void Info(string Format, params object?[] Args) => Verbose(LogVerbosity.Info, Format, Args);

    public static void Warning(string Format, params object?[] Args) => Verbose(LogVerbosity.Warning, Format, Args);

    public static void Error(string Format, params object?[] Args) => Verbose(LogVerbosity.Error, Format, Args);

    public static void Critical(string Format, params object?[] Args) => Verbose(LogVerbosity.Critical, Format, Args);

    public static Task FlushAsync(CancellationToken CToken = default)
    {
        if (StopToken.IsCancellationRequested)
        {
            return Task.CompletedTask;
        }

        TaskCompletionSource Source = new();
        Action(() => Source.SetResult());
        return Source.Task.WaitAsync(CToken);
    }

    public static void Shutdown()
    {
        lock (LogThread)
        {
            StopToken.Cancel();
            Monitor.Pulse(LogThread);
        }

        LogThread.Join();
    }

    private static string VerbosityToString(LogVerbosity verbosity) => verbosity switch
    {
        LogVerbosity.Trace => "dbug",
        LogVerbosity.Info => "info",
        LogVerbosity.Warning => "warn",
        LogVerbosity.Error => "erro",
        LogVerbosity.Critical => "ftal",
        _ => "????"
    };

    private static ConsoleColor VerbosityToColor(LogVerbosity verbosity) => verbosity switch
    {
        LogVerbosity.Trace => ConsoleColor.White,
        LogVerbosity.Info => ConsoleColor.DarkGreen,
        LogVerbosity.Warning => ConsoleColor.DarkYellow,
        LogVerbosity.Error => ConsoleColor.DarkRed,
        LogVerbosity.Critical => ConsoleColor.Red,
        _ => ConsoleColor.Red
    };

    private static void Run()
    {
        List<LogEntry> Entries = new();
        Thread.CurrentThread.Name = "[Log Thread]";
        ConsoleColor ForegroundColor = Console.ForegroundColor;

        while (true)
        {
            lock (LogThread)
            {
                while (!LogEntries.Any())
                {
                    Monitor.Wait(LogThread, 100);
                    if (StopToken.Token.IsCancellationRequested)
                    {
                        break;
                    }
                }

                if (StopToken.Token.IsCancellationRequested && LogEntries.Count == 0)
                {
                    break;
                }

                // Swap buffer.
                (Entries, LogEntries) = (LogEntries, Entries);
            }

            foreach (var Entry in Entries)
            {
                if (Entry.Message is null)
                {
                    Entry.LogAction!();
                }
                else
                {
                    Console.ForegroundColor = VerbosityToColor(Entry.Verbosity);
                    Console.Write(VerbosityToString(Entry.Verbosity));
                    Console.ForegroundColor = ForegroundColor;
                    Console.WriteLine(": {0}", Entry.Message);
                }
            }

            Entries.Clear();
        }
    }
}
